src/lib/protocols/http.c

```diff
@@ -886,54 +899,67 @@ static void check_content_type_and_change_protocol(struct ndpi_detection_module_
 	  flow->guessed_category = flow->category = ndpi_http_check_content(ndpi_struct, flow);
 	}
       }
+    }
+  }
 
-      if((flow->http.server == NULL) && (packet->server_line.len > 0)) {
-	int len = packet->server_line.len + 1;
-	
-	flow->http.server = ndpi_malloc(len);
-	if(flow->http.server) {
-	  strncpy(flow->http.server, (char*)packet->server_line.ptr,
-		  packet->server_line.len);
-	  flow->http.server[packet->server_line.len] = '\0';
-	}
+  /* check for host line (only if we don't already have an hostname) */
+  if(packet->host_line.ptr != NULL && flow->host_server_name[0] == '\0') {
+
+    NDPI_LOG_DBG2(ndpi_struct, "HOST line found %.*s\n",
+		  packet->host_line.len, packet->host_line.ptr);
+
+    /* Copy result for nDPI apps */
+    ndpi_hostname_sni_set(flow, packet->host_line.ptr, packet->host_line.len);
+
+    if(strlen(flow->host_server_name) > 0) {
+      char *double_col;
+      int a, b, c, d;
+
+      hostname_just_set = 1;
+
+      if(ndpi_is_valid_hostname(flow->host_server_name,
+				strlen(flow->host_server_name)) == 0) {
+	char str[128];
+
+	snprintf(str, sizeof(str), "Invalid host %s", flow->host_server_name);
+	ndpi_set_risk(ndpi_struct, flow, NDPI_INVALID_CHARACTERS, str);
+
+	/* This looks like an attack */
+	ndpi_set_risk(ndpi_struct, flow, NDPI_POSSIBLE_EXPLOIT, NULL);
       }
-    }
 
-    if(flow->http_detected && packet->content_line.ptr && *(char*)packet->content_line.ptr) {
-      /* Matching on Content-Type.
-          OCSP:  application/ocsp-request, application/ocsp-response
-       */
-      if(packet->content_line.len > 17 &&
-         strncmp((const char *)packet->content_line.ptr, "application/ocsp-", 17) == 0) {
-        NDPI_LOG_DBG2(ndpi_struct, "Found OCSP\n");
-        ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_OCSP, NDPI_PROTOCOL_HTTP, NDPI_CONFIDENCE_DPI);
+      double_col = strchr((char*)flow->host_server_name, ':');
+      if(double_col) double_col[0] = '\0';
+      if(ndpi_struct->packet.iph
+         && (sscanf(flow->host_server_name, "%d.%d.%d.%d", &a, &b, &c, &d) == 4)) {
```

Comment:
This format string does also accept invalid IPv4 addresses. Maybe add a IPv4 address validation method and set a risk if the address is invalid?

Reply:
We extract the address only to compare it with the address in the IP header to set the NDPI_HTTP_SUSPICIOUS_HEADER risk. An invalid address will trigger this risk even more so